### DISCUSSION TITLE
[@types/node] reports Hmac as deprecated. The class is not deprecated, only its direct instantiation is.

### DISCUSSION THREAD

**Discussion by SmashingQuasar** [DefinitelyTyped/DefinitelyTyped#70986] [Issues with a @types package]
Hey,

As of today, the following code will cause a deprecation error:

```ts
import { type Hmac, createHmac } from "node:crypto";

const myHmac: Hmac = createHmac("SHA256", "test");

console.debug(myHmac);
```

The `Hmac` class is marked as "deprecated". In fact, the class itself is not deprecated. Only directly instantiating it is deprecated.

![image](https://github.com/user-attachments/assets/270e8fe3-472d-4eba-938c-bc3e55a33e28)

This is a problem, as very strictly typed projects will automatically reject any code using deprecated members.
I understand the intention here, but I believe this is out of the scope of what typing can do in TypeScript. This is an issue that should most likely be handled by the Node.js team themselves.

If you really want to permanently address this issue, then I would recommend marking the constructor of `Hmac` as private. This is a harsher approach than marking `Hmac` as deprecated; however, it is semantically correct, whilst marking `Hmac` as deprecated is not.

Node.js could resolve this issue by simply throwing a deprecation warning in case the class is being directly instantiated without any call to `createHmac`.

Also, deprecations must be time-bound. This is not the case here, as `Hmac` will persist for the time being (most likely forever), which means `Hmac` will be a perpetual deprecation with this approach. This is an incorrect way of handling deprecations.

In any case, I think marking `Hmac` as deprecated is not the correct approach. Some things are unfortunately out of the scope of TypeScript, and this is one of them.

[a comment by an automated account is not shown]

**Comment by Renegade334** [comment]
The constructor is already marked as private, so the deprecation tag can just go.
